Metric spaces (#1162)
UniMath/agda-unimath: src/metric-spaces/functor-precategory-short-isometry-metric-spaces.lagda.md

> # The functor from the precategory of metric spaces and isometries to the precategory of metric spaces and short maps

Reviewer: This header is a bit too wordy

```suggestion
# The inclusion of isometries into the category of metric spaces and short maps
```

open import foundation.universe-levels

open import metric-spaces.isometry-metric-spaces
open import metric-spaces.precategory-of-metric-spaces-and-isometries
open import metric-spaces.precategory-of-metric-spaces-and-short-functions
open import metric-spaces.short-functions-metric-spaces
```

</details>

## Idea

Because every [isometry](metric-spaces.isometry-metric-spaces.md) between
[metric spaces](metric-spaces.metric-spaces.md) is also
[short](metric-spaces.short-functions-metric-spaces.md), there's a
[functor](category-theory.functors-precategories.md) between the
[precategory of metric spaces and isometries](metric-spaces.precategory-of-metric-spaces-and-isometries.md)
to the
[precategory of metric spaces and short maps](metric-spaces.precategory-of-metric-spaces-and-short-functions.md).

Moreover, since the map from isometry to short maps is an
[embedding](foundation.embeddings.md), this functor is
[faithful](category-theory.faithful-functors-precategories.md).

Finally, because short
[isomorphisms](category-theory.isomorphisms-in-precategories.md) are isometric,
this functor is also
[conservative](category-theory.conservative-functors-precategories.md).

> 

Reviewer: the functor is also an equivalence on objects

Author: > the functor is also an equivalence on objects

Do you mean this have a name and I should add it as a property of this functor? if so, how is it called? (sorry, my category theory classes were a long time ago and I can't find anything in the category theory module).

Reviewer: Well, there isn't a module for it, but it's a strong form of essential surjectivity of functors.

- [Essentially surjective functors](https://unimath.github.io/agda-unimath/category-theory.essentially-surjective-functors-precategories.html)
- [Split essentially surjective functors](https://unimath.github.io/agda-unimath/category-theory.split-essentially-surjective-functors-precategories.html)

Author: I added the `is-split-essentially-surjective` property. being the identity on objects, that was easy. (but the name is very long)

Reviewer: Sorry, I didn't explain myself properly. The property I was looking for is `is-equiv-obj-...` 
This is a stronger property than split essential surjectivity for precategories, but it is also conceptually related.

Reviewer: It is also a property of the functor, as opposed to split essential surjectivity which may be structure in general.

Reviewer: the full signature would be
```agda
is-equiv-obj-functor-precategory-short-isometry-Metric-Space :
  is-equiv (obj-functor-Precategory (functor-precategory-short-isometry-Metric-Space l1 l2))
```

Author: > Sorry, I didn't explain myself properly. The property I was looking for is `is-equiv-obj-...` This is a stronger property than split essential surjectivity for precategories, but it is also conceptually related.

oh. I see. well, I guess the `is-split-essentially-surjective-xxx` doesn't hurt (or do you want me to remove it?).

do you want `  is-id-obj-functor-precategory-short-isometry-Metric-Space` too? so basically 

### The precategory functor from isometries to short maps is the identity on objects

```agda
module _
  (l1 l2 : Level)
  where

  is-id-obj-functor-precategory-short-isometry-Metric-Space :
    ( obj-functor-Precategory
      ( precategory-isometry-Metric-Space)
      ( precategory-short-function-Metric-Space)
      ( functor-precategory-short-isometry-Metric-Space l1 l2)) ＝
    ( id)
  is-id-obj-functor-precategory-short-isometry-Metric-Space = refl
```

### The precategory functor from isometries to short maps is an equivalence on objects

```agda
module _
  (l1 l2 : Level)
  where

  is-equiv-obj-functor-precategory-short-isometry-Metric-Space :
    is-equiv
      ( obj-functor-Precategory
        ( precategory-isometry-Metric-Space)
        ( precategory-short-function-Metric-Space)
        ( functor-precategory-short-isometry-Metric-Space l1 l2))
  is-equiv-obj-functor-precategory-short-isometry-Metric-Space =
    is-equiv-id
```

Reviewer: > oh. I see. well, I guess the `is-split-essentially-surjective-xxx` doesn't hurt (or do you want me to remove it?).

It's redundant and not as useful, so I think it should be removed.

> do you want `is-id-obj-functor-precategory-short-isometry-Metric-Space` too? 

No.

Reviewer: Note that none of the below statements fully respect the principle of isomorphism invariance

- The functor is an equivalence on objects
- The functor is split essentially surjective
- The functor is the identity on objects

However, they are all useful because they allow you to conclude that you might have an isomorphism of precategories. The bottom line is sort of that the concept of "precategories" is evil, which is why it's not that useful to formalize all of the properties mentioned and it should suffice to prove one of them. Out of the mentioned properties, I think being an equivalence on objects is "the most" benign but you may be able to argue otherwise too, and either way you are negotiating with the devil.

Author: ok. thanks for the explanation.

```

### The precategory functor from isometries to short maps is conservative

```agda
module _
  (l1 l2 : Level)
  where

  is-conservative-functor-precategory-short-isometry-Metric-Space :
    is-conservative-functor-Precategory
      (precategory-isometry-Metric-Space)
      (precategory-short-function-Metric-Space)
      (functor-precategory-short-isometry-Metric-Space l1 l2)
  is-conservative-functor-precategory-short-isometry-Metric-Space {A} {B} f H =
    ( isometric-inv) ,
    ( eq-htpy-map-isometry-Metric-Space
      ( B)
      ( B)
      ( comp-isometry-Metric-Space B A B f isometric-inv)
      ( isometry-id-Metric-Space B)
      ( is-section-map-inv-is-equiv is-equiv-f)) ,
    ( eq-htpy-map-isometry-Metric-Space
      ( A)
      ( A)
      ( comp-isometry-Metric-Space A B A isometric-inv f)
      ( isometry-id-Metric-Space A)
      ( is-retraction-map-inv-is-equiv is-equiv-f))
    where

    is-equiv-f : is-equiv (map-isometry-Metric-Space A B f)
    is-equiv-f =
      is-equiv-is-iso-precategory-short-function-Metric-Space
        ( A)
        ( B)
        ( short-isometry-Metric-Space A B f)
        ( H)

    isometric-inv : isometry-Metric-Space B A

>     isometric-inv =

Reviewer: these need new names

Author: ok, I guess I'm abusing the `where` blocks. I'll refactor the proof to be more consistent with the guidelines.
